test/Test-Stores.shared.js

```diff
@@ -0,0 +1,180 @@
+    it('should reject when the file does not exist', function () {
+      const req = { file_id: '1234' }
+      return this.server.datastore.remove(req).should.be.rejected()
+    })
+
+    it('should delete the file when it does exist', async function () {
+      const fileDeletedEvent = sinon.fake()
+      const req = { headers: { 'upload-length': 1000 }, url: this.storePath }
+
+      this.server.datastore.on(EVENTS.EVENT_FILE_DELETED, fileDeletedEvent)
+
+      const file = await this.server.datastore.create(req)
+      await this.server.datastore.remove({ file_id: file.id })
+      assert.equal(fileDeletedEvent.calledOnce, true)
+    })
+  })
+}
+
+exports.shouldWriteUploads = function () {
+  describe('write', function () {
+    it('should reject write streams that cant be opened', function () {
+      const stream = fs.createReadStream(this.testFilePath)
+      return this.server.datastore.write(stream, null, 0).should.be.rejectedWith(500)
+    })
+
+    xit('should open a stream, resolve the new offset, and emit upload complete', function (done) {
+      const uploadCompleteEvent = sinon.fake()
+
+      this.server.datastore.on(EVENTS.EVENT_UPLOAD_COMPLETE, uploadCompleteEvent)
+
+      const stream = fs.createReadStream(this.testFilePath)
+      const name = this.testFileName
+      const size = this.testFileSize
+
+      stream.once('open', () => {
+        this.server.datastore
+          .write(stream, name, 0)
+          .then((offset) => {
+            assert.equal(offset, size)
+            assert.equal(uploadCompleteEvent.calledOnce, true)
+            return done()
+          })
+          .catch(done)
+      })
+    })
+
+    xit('should settle on closed input stream', function (done) {
+      const req = {
+        headers: {
+          'upload-length': this.testFileSize.toString(),
+          'upload-metadata': 'foo bar',
+        },
+        url: this.storePath,
+      }
+
+      const stream = fs.createReadStream(this.testFilePath)
+
+      stream.pause()
+      stream.on('data', () => stream.destroy())
+
+      this.server.datastore
+        .create(req)
+        .then((file) => {
+          return this.server.datastore.write(stream, file.id, 0)
+        })
+        .catch(() => {})
+        .finally(() => done())
+    })
+  })
+}
+
+exports.shouldHandleOffset = function () {
+  describe('getOffset', function () {
+    it('should reject non-existant files', function () {
+      return this.server.datastore.getOffset('doesnt_exist').should.be.rejectedWith(404)
+    })
+
+    it('should reject directories', function () {
+      return this.server.datastore.getOffset('').should.be.rejectedWith(404)
+    })
+
+    it('should resolve the stats for existant files', function () {
+      const req = { headers: { 'upload-length': this.testFileSize } }
+
+      this.server.datastore.create(req).then((file) => {
+        this.server.datastore.getOffset(file.id).should.be.fulfilledWith(this.testFileSize)
```

Comment:
I don't fully understand this test. We create an upload of the size `this.testFileSize` and then we obtain the upload offset without uploading anything. Shouldn't the offset by 0 instead of `this.testFileSize` in this case?

Reply:
AFAIK `create` already fully uploads the file so the offset is the file size.